### PR TITLE
add modal when specified KOTS version does not match current version

### PR DESCRIPTION
**Pull request by GraysonNull**
#### What type of PR is this?
type::feature

#### What this PR does / why we need it:
If a new version is trying to be deployed but the vendor has specified a KOTS version that is different than what is currently installed the user should see a message that they need to upgrade KOTS before proceeding with the new version.

#### Special notes for your reviewer:
Classic UI will an error "Incompatible KOTS version" and a 'Show details' link which will open this modal.
Screen Shot 2022-01-28 at 10 31 44 AM

The new UI has more space and can handle this error better without the need of a modal. Here is what the error looks like on the dashboard as well as on the app version history page
Screen Shot 2022-01-28 at 12 09 53 PM

Screen Shot 2022-01-28 at 12 08 50 PM

#### Does this PR introduce a user-facing change?
Yes
```release-note
Adds functionality to display a modal with commands for upgrading KOTS if there the version attempting to be deployed has a different KOTS version specified than the one the user has installed.
```

#### Does this PR require documentation?
NONE

### CODE REVIEW

**Comment by markpundsack** [comment]
As discussed, the instructions need to say something about the actual `kubectl kots admin-console upgrade` command, even if it has to leave the specific CLI flags up to the reader. We should be able to indicate the namespace. We should *not* output any passwords.

It might be interested to detect whether a registry was specified, and at least include the `kubectl kots admin-console push-images` command (again, without displaying the password).

As I understand, a user is either going to do a simple:
```
curl https://kots.io/install/version/<version> | bash
kubectl kots admin-console upgrade -n <namespace>
```
**or** the more complicated:
```
curl https://kots.io/install/version/<version> | bash
kubectl kots admin-console push-images ./kotsadm.tar.gz private.registry.host/application-name \
  --registry-username rw-username \
  --registry-password rw-password
kubectl kots admin-console upgrade \
  --kotsadm-registry private.registry.host/application-name \
  --registry-username ro-username \
  --registry-password ro-password \
  -n <namespace>
```

There are, of course, other flags that a user may have specified, like `--wait-duration` or `--context`. It's probably impossible for us to know if those are necessary. Even if we record it from the last install/upgrade, someone's kubectl context could be different.

File: web/src/components/apps/AppDetailPage.jsx [hunk 1]
```diff
@@ -464,28 +473,32 @@ class AppDetailPage extends Component {
...
-                onCopyText={<span className="u-textColor--success">Command has been copied to your clipboard</span>}
-              >
-                kubectl krew install kots
-                {`kubectl kots download --namespace ${this.props.appNameSpace} --slug ${this.props.match.params.slug}`}
-              </CodeSnippet>
+              <h2 className="u-fontSize--largest u-textColor--primary u-fontWeight--bold u-lineHeight--normal">You must update KOTS to deploy this version</h2>
+              <p className="u-fontSize--normal u-textColor--bodyCopy u-lineHeight--normal u-marginBottom--20">This version of {app?.name} requires a version of KOTS that is different from what you currently have installed. Follow the steps below to upgrade KOTS to the required version.</p>
+              {app?.isAirgap ?
+                <p className="u-fontSize--normal u-textColor--bodyCopy u-lineHeight--normal u-marginBottom--20">To get the version of KOTS specified in this release ({requiredKotsUpdateObj?.incompatibleKotsVersion}) you can visit the download portal or talk to your vendor. Once you have upgraded to the specified version you will be abel to deploy this version.</p>
```
Comment (hkosova) [inline web/src/components/apps/AppDetailPage.jsx:489]: Typo: abel -> able
```suggestion
                <p className="u-fontSize--normal u-textColor--bodyCopy u-lineHeight--normal u-marginBottom--20">To get the version of KOTS specified in this release ({requiredKotsUpdateObj?.incompatibleKotsVersion}) you can visit the download portal or talk to your vendor. Once you have upgraded to the specified version you will be able to deploy this version.</p>
```

Comment (GraysonNull) [reply web/src/components/apps/AppDetailPage.jsx]: message is different and coming from the API now

**Comment by markpundsack** [comment]
@GraysonNull Can you update the first comment with new screenshots?

File: web/src/components/apps/AppVersionHistory.jsx [hunk 2]
```diff
@@ -322,42 +322,6 @@ class AppVersionHistory extends Component {
         </div>
       );
     }
-    // if (hasDiffSummaryError) {
```
Comment (sgalsaleh) [inline web/src/components/apps/AppVersionHistory.jsx:325]: cleanup?

File: web/src/components/apps/DashboardClassic.jsx [hunk 3]
```diff
@@ -298,11 +298,22 @@ class DashboardClassic extends Component {
           if (response.status !== "running" && !this.props.isBundleUploading) {
             this.state.updateChecker.stop();
 
-            this.setState({
-              checkingForUpdates: false,
-              checkingUpdateMessage: response.currentMessage,
-              checkingForUpdateError: response.status === "failed"
-            });
+            if (response.status === "failed") {
+              if (response.currentMessage.includes("Upgrade KOTS to version")) {
```
Comment (sgalsaleh) [inline web/src/components/apps/DashboardClassic.jsx:302]: maybe just
```suggestion
              if (response.currentMessage.includes("Upgrade KOTS")) {
```

File: web/src/components/apps/AppVersionHistoryClassic.jsx [hunk 4]
```diff
@@ -501,11 +501,22 @@ class AppVersionHistoryClassic extends Component {
           if (response.status !== "running" && !this.props.isBundleUploading) {
             this.state.updateChecker.stop();
 
-            this.setState({
-              checkingForUpdates: false,
-              checkingUpdateMessage: response.currentMessage,
-              checkingForUpdateError: response.status === "failed"
-            });
+            if (response.status === "failed") {
+              if (response.currentMessage.includes("Upgrade KOTS to version")) {
```
Comment (sgalsaleh) [inline web/src/components/apps/AppVersionHistoryClassic.jsx:505]: same here

File: web/src/components/apps/AppVersionHistory.jsx [hunk 5]
```diff
@@ -602,11 +566,22 @@ class AppVersionHistory extends Component {
           if (response.status !== "running" && !this.props.isBundleUploading) {
             this.state.updateChecker.stop();
 
-            this.setState({
-              checkingForUpdates: false,
-              checkingUpdateMessage: response.currentMessage,
-              checkingForUpdateError: response.status === "failed"
-            });
+            if (response.status === "failed") {
+              if (response.currentMessage.includes("Upgrade KOTS to version")) {
```
Comment (sgalsaleh) [inline web/src/components/apps/AppVersionHistory.jsx:570]: same here

Comment (sgalsaleh) [inline web/src/components/apps/AppVersionHistory.jsx:570]: isn't this the new ui? we don't need this, right?

File: web/src/components/apps/AppVersionHistory.jsx (hunk 2 above)
Comment (GraysonNull) [reply web/src/components/apps/AppVersionHistory.jsx:325]: yeah

**Review by sgalsaleh — DISMISSED** [review]

**Review by sgalsaleh — APPROVED** [review]

**Pull request merged**